knodle/transformation/rule_reduction.py

```diff
@@ -0,0 +1,168 @@
+import logging
+from typing import Tuple, Dict, Iterable
+
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+
+def reduce_lf_matches(
+    rule_matches_z: np.ndarray, mapping_rule_class_t: np.ndarray, drop_rules: bool = False,
+    max_rules: int = None, min_coverage: float = None, rule_matches_rest: Dict[str, np.ndarray] = None
+) -> Dict[str, np.ndarray]:
+
+    """
+    This is the main function to be used for reduction of rules.
+    It finds the less effective rules by coverage in the train rule_matches_z and reduces them in the match matrix.
+    The reduction is either done by dropping the rules (if drop_rules is True), or by creating one merged rule per label
+    from the excessive rules selected for reduction (drop_rules is False).
+    The mapping_rule_class_t matrix is adjusted accordingly. At least one of the criteria [max_rules, min_coverage]
+    should be provided.
+    Validation and test matches can be provided in rule_matches_rest to be reduced in the same manner as training data.
+
+    Args:
+        rule_matches_z: main match matrix, based on which the LFs to keep are selected
+        drop_rules: If True, the rules will be completely discarded, otherwise they will be
+            represented as one rule (column) in reduced match matrix.
+        max_rules: maximal number of rules to keep as-is. If drop_rules=False, with additional merged rules the total
+            number of rules in the output may exceed this limit by #classes at most.
+        min_coverage: minimal coverage required for a rule to be kept
+        rule_matches_rest: additional rule_matches (dev, test) which should be reduced in the same manner
+    Returns: Dictionary with "rule_matches_z", "mapping_rule_class_t" and keys from rule_matches_rest if provided.
+    """
+    if max_rules is None and min_coverage is None:
+        logger.info("No filtering criteria ('max_lf' or 'min_coverage' for LF specified, "
+                    "returning the original LF matches.")
+        return {"rule_matches_z": rule_matches_z, "mapping_rule_class_t": mapping_rule_class_t}
+
+    coverage_per_rule = rule_matches_z.sum(0) / rule_matches_z.shape[0]
+
+    # create mask to indicate which rules will be kept unchanged
+    lf_kept_mask = np.zeros(mapping_rule_class_t.shape[0], dtype=np.bool)
+
+    # take top N rules
+    if max_rules is not None:
```

Comment:
to be sure: is it a valid use case when user wants both filtering by max rules and coverage?